### PR TITLE
[SPARK-10392] [SQL] Pyspark - Wrong DateType support on JDBC connection

### PR DESCRIPTION
**Pull request by 0x0FFF**
This PR addresses issue [SPARK-10392](https://issues.apache.org/jira/browse/SPARK-10392)
The problem is that for "start of epoch" date (01 Jan 1970) PySpark class DateType returns 0 instead of the `datetime.date` due to implementation of its return statement

Issue reproduction on master:

```
>>> from pyspark.sql.types import *
>>> a = DateType()
>>> a.fromInternal(0)
0
>>> a.fromInternal(1)
datetime.date(1970, 1, 2)
```

### CODE REVIEW

**Comment by JoshRosen** [comment]
Jenkins, this is ok to test.

**Comment by andrewor14** [comment]
@davies

File: python/pyspark/sql/types.py [hunk 1]
```diff
@@ -168,10 +168,12 @@ def needConversion(self):
         return True
 
     def toInternal(self, d):
-        return d and d.toordinal() - self.EPOCH_ORDINAL
+        if d is not None:
```
Comment (davies) [inline python/pyspark/sql/types.py]: I think this change is not needed, `d` will always be true if it's not None

**Comment by davies** [comment]
@0x0FFF Thanks for working on this, could you add a regression test for it?

File: python/pyspark/sql/types.py (hunk 1 above)
Comment (0x0FFF) [inline python/pyspark/sql/types.py]: It will be false for None, empty string, empty list, 0. This way it will be returned instead of the date:
```
>>> [] and 'foo'
[]
>>> 0 and 'foo'
0
>>> 1 and 'foo'
'foo'
```

Comment (davies) [inline python/pyspark/sql/types.py]: `d` could not be list/int/string, it can only be Date or None.

**Comment by 0x0FFF** [comment]
Added regression test

File: python/pyspark/sql/types.py (hunk 1 above)
Comment (0x0FFF) [inline python/pyspark/sql/types.py]: Sorry, I see. Should I revert this change? It is mostly for both methods to follow the same logic, as `return x and y` is not a very readable code in my opinion. Also the class `TimestampType` is implemented in a similar way - both methods are starting with input parameter check for None

Comment (davies) [inline python/pyspark/sql/types.py]: Either is good to me.

File: python/pyspark/sql/tests.py [hunk 2]
```diff
@@ -1164,6 +1164,9 @@ def test_window_functions_without_partitionBy(self):
         for r, ex in zip(rs, expected):
             self.assertEqual(tuple(r), ex[:len(r)])
 
+    def test_datetype_equal_zero(self):
```
Comment (davies) [inline python/pyspark/sql/tests.py]: This test case don't need sqlCtx, it's better to be inside DataTypeTests.

**Comment by 0x0FFF** [comment]
Moved regression test to DataTypeTests class

**Comment by davies** [comment]
LGTM, will merge once it pass tests.

**Comment by 0x0FFF** [comment]
Failed mllib test for python2.6, I didn't change anything that might have affected it. Same test passes locally on my machine

**Comment by 0x0FFF** [comment]
Jenkins, retest this please.

**Comment by davies** [comment]
@0x0FFF The JVM died during tests.

**Comment by davies** [comment]
merged into master, thanks!
